Mathlib/Tactic/Widget/StringDiagram.lean

```diff
@@ -0,0 +1,608 @@
+/-
+Copyright (c) 2024 Yuma Mizuno. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yuma Mizuno
+-/
+import ProofWidgets.Component.PenroseDiagram
+import ProofWidgets.Presentation.Expr
+import Mathlib.CategoryTheory.Monoidal.Category
+
+
+/-!
+# String Diagrams
+
+This file provides tactic/meta infrastructure for displaying string diagrams for morphisms
+in monoidal categories in the infoview.
+
+-/
+
+namespace Mathlib.Tactic.Widget.StringDiagram
+
+open Lean Meta Elab
+open CategoryTheory
+
+/-- Expressions for atomic 1-morphisms. -/
+structure Atom₁ : Type where
+  /-- Extract a Lean expression from an `Atom₁` expression. -/
+  e : Expr
+
+/-- Expressions for 1-morphisms. -/
+inductive Mor₁ : Type
+  /-- Construct the expression for `𝟙_ C`. -/
+  | id : Expr → Mor₁
+  /-- Construct the expression for `X ⊗ Y` -/
+  | comp : Mor₁ → Mor₁ → Mor₁
+  /-- Construct the expression for an atomic 1-morphism. -/
+  | of : Atom₁ → Mor₁
+  deriving Inhabited
+
+/-- Extract a Lean expression from a `Mor₁` expression. -/
+def Mor₁.e : Mor₁ → MetaM Expr
+  | .id C => do
+    mkAppOptM ``MonoidalCategoryStruct.tensorUnit #[C]
+  | .comp f g => do
+    mkAppM ``MonoidalCategoryStruct.tensorObj #[← Mor₁.e f, ← Mor₁.e g]
+  | .of f => return f.e
+
+/-- Converts a 1-morphism into a list of its underlying expressions. -/
+def Mor₁.toList : Mor₁ → List Expr
+  | .id _ => []
+  | .comp f g => f.toList ++ g.toList
+  | .of f => [f.e]
+
+/-- Construct a `Mor₁` expression from a Lean expression. -/
+partial def toMor₁ (e : Expr) : Mor₁ :=
+  match e.getAppFnArgs with
+  | (``MonoidalCategoryStruct.tensorUnit, #[C, _, _]) => Mor₁.id C
+  | (``MonoidalCategoryStruct.tensorObj, #[_, _, _, f, g]) => (toMor₁ f).comp (toMor₁ g)
+  | _ => Mor₁.of ⟨e⟩
+
+/-- Expressions for atomic structural 2-morphisms. -/
+inductive StructuralAtom : Type
+  /-- Construct the expression for the associator. -/
+  | associator (f g h : Mor₁) : StructuralAtom
+  /-- Construct the expression for the inverse of the associator. -/
+  | associatorInv (f g h : Mor₁) : StructuralAtom
+  /-- Construct the expression for the left unitor. -/
+  | leftUnitor (f : Mor₁) : StructuralAtom
+  /-- Construct the expression for the inverse of the left unitor. -/
+  | leftUnitorInv (f : Mor₁) : StructuralAtom
+  /-- Construct the expression for the right unitor. -/
+  | rightUnitor (f : Mor₁) : StructuralAtom
+  /-- Construct the expression for the inverse of the right unitor. -/
+  | rightUnitorInv (f : Mor₁) : StructuralAtom
```

Comment:
Can you give the Lean expression for each case in the comment, like you did above?